### ISSUE TITLE
Che fails to communicate with Workspace agent

### ISSUE THREAD

**Issue by itaimalek** [eclipse-che/che#973]
Hi,
I'm have eclipse che running natively on a CentOS 7 instance at google cloud platform.
I'm able to get che interface on my browser, but cannot start a project.

**from my che.properties file**

```
machine.docker.local_node_host=172.17.0.1
api.endpoint=http://_my instance external IP_:${SERVER_PORT}/ide/api
```

**from my /etc/hosts** (putted manualy):
`172.17.0.1  che-host`

**installed Docker as pre-requirement**
`Docker version 1.10.3, build 20f81dd`

**from my ~/.bash_profile file**

```
export JAVA_HOME=/usr/lib/jvm/jre-1.8.0-openjdk-1.8.0.77-0.b03.el7_2.x86_64
export CHE_HOME=/opt/eclipse-che-4.1.0
export CHE_LOCAL_CONF_DIR=$CHE_HOME/conf
export CHE_DOCKER_MACHINE_HOST=172.17.0.1
```

**ifconfig output**

```
docker0: flags=4163<UP,BROADCAST,RUNNING,MULTICAST>  mtu 1500
        inet 172.17.0.1  netmask 255.255.0.0  broadcast 0.0.0.0
        inet6 fe80::42:68ff:fe05:66b1  prefixlen 64  scopeid 0x20<link>
        ether 02:42:68:05:66:b1  txqueuelen 0  (Ethernet)
        RX packets 3025  bytes 334688 (326.8 KiB)
        RX errors 0  dropped 0  overruns 0  frame 0
        TX packets 2563  bytes 390278 (381.1 KiB)
        TX errors 0  dropped 0 overruns 0  carrier 0  collisions 0

eth0: flags=4163<UP,BROADCAST,RUNNING,MULTICAST>  mtu 1460
        inet <google internal network IP for instance>  netmask 255.255.255.255  broadcast 10.132.0.2
        inet6 fe80::4001:aff:fe84:2  prefixlen 64  scopeid 0x20<link>
        ether 42:01:0a:84:00:02  txqueuelen 1000  (Ethernet)
        RX packets 402499  bytes 1381254390 (1.2 GiB)
        RX errors 0  dropped 0  overruns 0  frame 0
        TX packets 292323  bytes 58381431 (55.6 MiB)
        TX errors 0  dropped 0 overruns 0  carrier 0  collisions 0

lo: flags=73<UP,LOOPBACK,RUNNING>  mtu 65536
        inet 127.0.0.1  netmask 255.0.0.0
        inet6 ::1  prefixlen 128  scopeid 0x10<host>
        loop  txqueuelen 0  (Local Loopback)
        RX packets 1061  bytes 57399 (56.0 KiB)
        RX errors 0  dropped 0  overruns 0  frame 0
        TX packets 1061  bytes 57399 (56.0 KiB)
        TX errors 0  dropped 0 overruns 0  carrier 0  collisions 0

veth53e0def: flags=4163<UP,BROADCAST,RUNNING,MULTICAST>  mtu 1500
        inet6 fe80::1005:77ff:fe1d:83a4  prefixlen 64  scopeid 0x20<link>
        ether 12:05:77:1d:83:a4  txqueuelen 0  (Ethernet)
        RX packets 42  bytes 6574 (6.4 KiB)
        RX errors 0  dropped 0  overruns 0  frame 0
        TX packets 40  bytes 7991 (7.8 KiB)
        TX errors 0  dropped 0 overruns 0  carrier 0  collisions 0
```

**firewalld rules for trusted zone**

```
instance-1 ~]$ sudo firewall-cmd --permanent --zone=trusted --list-all
trusted (default, active)
  interfaces: docker0
  sources:
  services:
  ports: 32768-65535/tcp 32768-65535/udp 4243/tcp 4401/tcp 4411/tcp 4403/tcp 22/tcp
  masquerade: no
  forward-ports:
  icmp-blocks:
  rich rules:

```

**output from che console in browser**

```
[STDOUT] 2016-04-05 19:40:39,511[main] [INFO] [o.a.c.http11.Http11NioProtocol 470] - Starting ProtocolHandler ["http-nio-4401"]
[STDOUT] 2016-04-05 19:40:39,515[main] [INFO] [o.a.catalina.startup.Catalina 642] - Server startup in 4795 ms
[STDOUT] 2016-04-05 19:40:39,515[main] [INFO] [o.a.catalina.startup.Catalina 642] - Server startup in 4795 ms
[STDOUT] 2016-04-05 19:40:39,712[nio-4401-exec-1] [WARN ] [i.ContainerResponse$BodyWriter 90] - Client has aborted connection. Response writing omitted.
[STDOUT] 2016-04-05 19:40:39,712[nio-4401-exec-1] [WARN ] [i.ContainerResponse$BodyWriter 90] - Client has aborted connection. Response writing omitted.
```

this happens in Starting workspace agent stage, after below stages 

```
Initializing workspace ... completed
Starting workspace runtime ... completed
```

**Comment by ghost** [comment]
The problem here is that browser tries to connect to a workspace agent and it uses `CHE_DOCKER_MACHINE_HOST` which is also used by the server to ping ws-agent.

So, `CHE_DOCKER_MACHINE_HOST` should be an externally available IP.

**Comment by itaimalek** [comment]
Thanks for reply,
Does that say that `docker0` interface should get a different/same **external** IP?

**Comment by ghost** [comment]
In your case it should be an external instance IP. This IP is used twice:
1. Server ping Agent - it was success in your case
2. Browser connects to ws-agent (ws://) - failure. You may open browser's dev console, network tab and take a look as websocket connections initiated.

**Comment by itaimalek** [comment]
ok, so the same IP I get for external connection is the IP to configure as `CHE_DOCKER_MACHINE_HOST` regardless to the fact that `docker0` is set on `172.17.0.1`
correct?
is it the same for `machine.docker.local_node_host` also?

**Comment by ghost** [comment]
Yes, `docker0` IP is used for ws-agent to ping server. It is a different property - `machine.docker.che_api.endpoin`

**Comment by itaimalek** [comment]
so, to sum things up
`CHE_DOCKER_MACHINE_HOST`=external IP
`machine.docker.local_node_host`= external IP
`machine.docker.che_api.endpoint`=`docker0` IP
correct?

**Comment by ghost** [comment]
Not sure this is needed at all - `machine.docker.local_node_host= external IP`

**Comment by itaimalek** [comment]
ok, I set everything as discussed above, now I get a timeout right after this:
`[STDOUT] 2016-04-06 09:51:35,106[main] [INFO] [o.a.catalina.startup.Catalina 642] - Server startup in 4977 ms`

with this message:
`Error when starting agent
Unable to start workspace agent. Error when trying to start the workspace agent: Timeout reached. The Che server has been unable to verify that your workspace's agent has successfully booted. Either the workspace is unreachable, the agent had an error during startup, or your workspace is starting slowly. You can configure machine.ws_agent.max_start_time_ms in Che properties to increase the timeout.`

Could be a FW issue?

**Comment by ghost** [comment]
Take a look at websocket connections when running a workspace. If the workspace has started, and you can actually see a container, then it's browser that fails to connect to it. Is the ephemeral port range opened?

**Comment by itaimalek** [comment]
sorry for the nube question, what is the ws:// address?

**Comment by ghost** [comment]
Open Chrome dev console, Network tab. WS checkbox. Look at the address `ws://IP_COMES_HERE` if they are pending forever, it means ports are closed.

**Comment by itaimalek** [comment]
![image](https://cloud.githubusercontent.com/assets/8522568/14314793/2117249a-fc02-11e5-93ce-f1f958ffc1ff.png)

This is what I'm getting when I click "Open in IDE"

**my settings are now**:
`machine.docker.che_api.endpoint=http://172.17.0.1:8080/ide/api`
`machine.docker.local_node_host=NULL`
`api.endpoint=http://<external-IP>:${SERVER_PORT}/ide/api`

**~/.bash_profile**

```

export JAVA_HOME=/usr/lib/jvm/jre-1.8.0-openjdk-1.8.0.77-0.b03.el7_2.x86_64
export CHE_HOME=/opt/eclipse-che-4.1.0
export CHE_LOCAL_CONF_DIR=$CHE_HOME/conf
export CHE_DOCKER_MACHINE_HOST="<external -IP>"

```

**my gcloud access rule for ports**

```
IPProtocol: TCP
creationTimestamp: '2016-04-05T12:34:02.172-07:00'
description: ''
id: '5658315597393014213'
kind: compute#forwardingRule
name: docker-che
portRange: 32768-65535
```

```
IPProtocol: TCP
creationTimestamp: '2016-04-05T13:25:32.971-07:00'
description: ''
id: '7578762094395892179'
kind: compute#forwardingRule
name: docker-che1
portRange: 4401-4401
```

```
IPProtocol: TCP
creationTimestamp: '2016-04-05T13:26:18.086-07:00'
description: ''
id: '5395434655093203333'
kind: compute#forwardingRule
name: docker-che2
portRange: 4403-4403
```

```
IPProtocol: TCP
creationTimestamp: '2016-04-05T13:27:03.670-07:00'
description: ''
id: '4333604987313168760'
kind: compute#forwardingRule
name: docker-che3
portRange: 4411-4411
```

```
IPProtocol: TCP
creationTimestamp: '2016-04-05T04:23:47.691-07:00'
description: ''
id: '733884641885167276'
kind: compute#forwardingRule
name: eclipse-che
portRange: 8080-8080
```

I'va also turned off firewalld service in my instance.

What am I missing?

**Comment by ghost** [comment]
What is the URL of the websocket? Can you hover on it?

**Comment by itaimalek** [comment]
![image](https://cloud.githubusercontent.com/assets/8522568/14318035/abab2a00-fc13-11e5-9f2a-6f3e3300aa9e.png)

sorry for delay - I had a meeting
The IP is the external IP for my instance.

**Comment by ghost** [comment]
Can you confirm the ephemeral port range is opened? Also, can you share server logs? If server can ping agent but your browser cannot, you should make sure ports are opened.

**Comment by itaimalek** [comment]
I created a brand new instance in gCloud , this time with port-forwarding enabled, I'm installing all from scratch again,
will update once done.

**Comment by itaimalek** [comment]
I'm happy to say the issue was finally resolved, the problem was with opening the firewall ports in the gcloud network assigned to my instances, beside of an instance FW there is a  virtual network FW that gcloud provides.
using:
`gcloud compute firewall-rules create docker-range --allow tcp:32768-65535`
there is an option to set IP restriction as well.

Thanks eivantsov -I learned allot from this thread.
excited to start using Che 
keep up the good work !
